common/workunit/workunit.cpp

```diff
@@ -2487,8 +2487,74 @@ class CompoundAggregator : implements StatisticAggregator
     IArrayOf<StatisticAggregator> aggregators;
 };
 
+// Conditionally aggregate costs by
+// * excluding hThor costs when excludehThor=true
+// * or hThor costs that are children of workflow that already has a cost
+class FilteredCostAggregator : public CInterfaceOf<IWuScopeVisitor>
+{
+public:
+    FilteredCostAggregator(bool _excludeHThor) : excludeHThor(_excludeHThor) {}
 
-//To calculate aggregates, create a scope iterator, and an instance of a StatisticAggregator, and play the attributes through the interface
+    virtual void noteStatistic(StatisticKind kind, unsigned __int64 value, IConstWUStatistic & extra) override
+    {
+        if (extra.getCreatorType() == SCThthor)
+        {
+            if (excludeHThor)
+                return;
+            if (haveWorkFlowCost)
+            {
+                // Exclude hThor costs that are children of workflow that already has a cost
+                if (isParentScope(workflowScope, extra.queryScope()))
+                    return;
+                else
+                    haveWorkFlowCost = false;
+            }
+        }
+        if (value && extra.getScopeType()==SSTworkflow)
+        {
+            haveWorkFlowCost = true;
+            workflowScope.set(extra.queryScope());
+        }
+        totalCost += value;
+    }
+    virtual void noteAttribute(WuAttr attr, const char * value) override { throwUnexpected(); }
+    virtual void noteHint(const char * kind, const char * value) override { throwUnexpected(); }
+    virtual void noteException(IConstWUException & exception) { throwUnexpected(); }
+    virtual cost_type getTotalCost() const { return totalCost; }
+protected:
+    const bool excludeHThor;
+    bool haveWorkFlowCost = false;
+    StringBuffer workflowScope;
+    cost_type totalCost = 0;
+};
+
+// Aggregrate Thor or hThor costs
+cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool excludeHThor)
```

Comment:
I think you are using this for 3 different scenarios:

a) Give me the total time so far for the job
b) What is the total time within a workflow item that was not executed locally (EclAgentWorkflowMachine::noteTiming)
c) How much time has been spent in this process that is not included in the current time.  (EclAgent::doProcess)

I think you might need 3 different approaches
a) Iterate all scopes.  Whenever you find a stat in a scope call nextSibling(), otherwise you will double count workflow items and the graphs within them
b) In three minds!
  1) Use the code is currently implemented - but with a bit more work on the filter to ensure subgraph stats are not iterated.
  2) If the target cluster type is "Thor" then Total the costs within a wfid, and add them to the wfid cost.
  3) Keep track of the time spent executing graphs within a workflow item, always add the cost for graphs, but subtract off the time for local graphs.
It might help to consider how it is affected if other items within the workflow start recording costs (e.g., sprays?).  I suspect (1) is the best option but it should document how the extra cases are handled.
c)  This should probably be done by calculating it based on the local execute time for the eclagent - i.e. subtracting off the time spent processing workflow items, but adding on the costs for all child items.  But it raises the question of how jobs that use the scheduler are costed.  Does it mean there is no cost for the workunit until it is complete.  What if it never completes?  Where is the costs for waiting for a persist to be rebuilt stored?

The PR should really contain a comment describing the how the costs are meant to work.  E.g., the costs at each level include the costs from all lower levels.

Reply:
In reply to 1: I'm not sure that it is iterating subgraph unnecessary.  I've emailed you some log extracts.